Let it snow

brettpenzer123 opened the pull request:
Well here goes...

thenamelessone1 commented:
what is this above here?
i don't understand

brettpenzer123 commented:
It is the snow eventually. Just awaiting further input by @Gymnasiast and hoping someone that can fix the mac issue.

At least I've been able to fix majority of the issues I've made during the end of that. I was using notepad++ to fix some of the code issues. Should have just opened it back in visual studio 👍

Broxzier commented on data/language/en-GB.txt:
```diff
@@ -3668,6 +3668,9 @@ STR_6346    :Tile Inspector: Decrease element height
 STR_6349    :Random title sequence
+STR_6350    :Snow
+STR_6351	:Heavy Snow
+STR_6352	:Blizzard
```
Use 4 spaces instead of tabs

Broxzier commented on src/openrct2-ui/windows/Cheats.cpp:
```diff
@@ -728,14 +731,14 @@ static void window_cheats_misc_mousedown(rct_window* w, rct_widgetindex widgetIn
             dropdownWidget = widget - 1;
 
-            for (i = 0; i < 6; i++)
+            for (i = 0; i < 9; i++)
```
This seems like a good opportunity to use `std::size(WeatherTypes)`

Broxzier commented on src/openrct2-ui/windows/Cheats.cpp:
```diff
@@ -728,14 +731,14 @@ static void window_cheats_misc_mousedown(rct_window* w, rct_widgetindex widgetIn
             window_dropdown_show_text_custom_width(
                 w->x + dropdownWidget->left, w->y + dropdownWidget->top, dropdownWidget->bottom - dropdownWidget->top + 1,
-                w->colours[1], 0, DROPDOWN_FLAG_STAY_OPEN, 6, dropdownWidget->right - dropdownWidget->left - 3);
+                w->colours[1], 0, DROPDOWN_FLAG_STAY_OPEN, 9, dropdownWidget->right - dropdownWidget->left - 3);
```
Same as previous note.

Broxzier commented on src/openrct2/drawing/Snow.cpp:
```diff
@@ -0,0 +1,99 @@
+static void DrawHeavySnow(IRainDrawer* rainDrawer, int32_t left, int32_t top, int32_t width, int32_t height);
+
+const DrawSnowFunc DrawSnowFunctions[] = {
+    nullptr,
```
I assume this exists for when you _don't_ want to draw snow, but wouldn't it be better to not call the DrawXSnow functions at all in these cases, instead of relying on this nullptr value?

brettpenzer123 replied:
DrawRainFunctions also has the same nullptr value and since snow is running alongside it using IRainDrawer it just made sense to keep it.

```
const DrawRainFunc DrawRainFunctions[] = {
    nullptr,
    &DrawLightRain,
    &DrawHeavyRain,
}; 
```

Snow is also using RAIN_LEVEL_LIGHT and RAIN_LEVEL_HEAVY. 
```
const WeatherState ClimateWeatherData[9] = {
    { 10, WEATHER_EFFECT_NONE, 0, RAIN_LEVEL_NONE, SPR_WEATHER_SUN },         // Sunny
    { 5, WEATHER_EFFECT_NONE, 0, RAIN_LEVEL_NONE, SPR_WEATHER_SUN_CLOUD },    // Partially Cloudy
    { 0, WEATHER_EFFECT_NONE, 0, RAIN_LEVEL_NONE, SPR_WEATHER_CLOUD },        // Cloudy
    { -2, WEATHER_EFFECT_RAIN, 1, RAIN_LEVEL_LIGHT, SPR_WEATHER_LIGHT_RAIN }, // Rain
    { -4, WEATHER_EFFECT_RAIN, 2, RAIN_LEVEL_HEAVY, SPR_WEATHER_HEAVY_RAIN }, // Heavy Rain
    { 2, WEATHER_EFFECT_STORM, 2, RAIN_LEVEL_HEAVY, SPR_WEATHER_STORM },      // Thunderstorm
    { -10, WEATHER_EFFECT_SNOW, 1, RAIN_LEVEL_LIGHT, SPR_WEATHER_SNOW },      // Snow
    { -15, WEATHER_EFFECT_SNOW, 2, RAIN_LEVEL_HEAVY, SPR_WEATHER_SNOW },      // Heavy Snow
    { -20, WEATHER_EFFECT_BLIZZARD, 2, RAIN_LEVEL_HEAVY, SPR_WEATHER_SNOW },  // Blizzard
};
```

```
        uint32_t rainType = gClimateCurrent.RainLevel;
        if (rainType != RAIN_LEVEL_NONE && !gTrackDesignSaveMode && !(viewFlags & VIEWPORT_FLAG_HIGHLIGHT_PATH_ISSUES))
        {
            auto drawFunc = DrawSnowFunctions[rainType];

```
I guess this just saves a lot of duplication in terms of copying most of the pre-existing rain drawing code.

Broxzier commented on src/openrct2/libopenrct2.vcxproj:
```diff
@@ -65,4 +65,4 @@
   </ItemGroup>
   <Import Project="$(VCTargetsPath)\Microsoft.Cpp.targets" />
-</Project>
+</Project>
```
Accidentally removed the EOL here.

Gymnasiast commented:
I'm mostly concerned with how it works and not necessarily the technical side of things - I will mostly leave that to @janisozaur and @ZehMatt .

Having said that, here are my thoughts about how it should work in my opinion:
I think there should be only one type of snow (just regular, light snowfall) and that people should not regard it as rain. In heavy snow or a blizzard, a park would be closed anyway. I also think it should only happen in cold climates (or at most in cold and cool-and-wet climates).

brettpenzer123 commented:
> I'm mostly concerned with how it works and not necessarily the technical side of things - I will mostly leave that to @janisozaur and @ZehMatt .
> 
> Having said that, here are my thoughts about how it should work in my opinion:
> I think there should be only one type of snow (just regular, light snowfall) and that people should not regard it as rain. In heavy snow or a blizzard, a park would be closed anyway. I also think it should only happen in cold climates (or at most in cold and cool-and-wet climates).

As of now, the blizzard is never picked from the climate since there isn't any difference from heavy snow and blizzard. Might even be worth removing it but someone else might be able to put some effect in there later on.

I think that heavy snow should stay just for the fact that during heavy rain (long periods) and thunderstorms a park would also close. I don't think RCT2 ever took realism into account with the weather. 

I did see a user come up with the park closing during night time but since the day/night cycle is way too quick and it was closed more than open. Since the weather at this point is unpredictable, we wouldn't know how long a park is closed and this could cause issues with all the peeps leaving from larger parks.

That is true, snow and heavy snow only show in cold and cold-and-wet climates.

I agree that peeps should react to the snow, even if it's just getting a hot drink to stay warm.

Gymnasiast commented:
Well, let's hear what other team members think about the heavy snow before we make a decision, then.

> I agree that peeps should react to the snow, even if it's just getting a hot drink to stay warm.

This behaviour is already implemented. (Conversely, cold food/drink does very well in hot weather.)

brettpenzer123 commented:
> Well, let's hear what other team members think about the heavy snow before we make a decision, then.
> 
> > I agree that peeps should react to the snow, even if it's just getting a hot drink to stay warm.
> 
> This behaviour is already implemented. (Conversely, cold food/drink does very well in hot weather.)

Isn't it temperature-related rather than weather?

ZehMatt commented on src/openrct2/drawing/X8DrawingEngine.cpp:
```diff
@@ -47,7 +49,15 @@ void X8RainDrawer::SetDPI(rct_drawpixelinfo* dpi)
 {
-    const uint8_t* pattern = RainPattern;
+    const uint8_t* pattern;
+    if (climate_is_snowing())
```
This should have nothing to do with the rain drawing code, best to move all out into its own class.

ZehMatt commented on src/openrct2/ui/UiContext.h:
```diff
@@ -29,6 +29,8 @@ namespace OpenRCT2
         interface IRainDrawer;
         using DrawRainFunc = void (*)(
             OpenRCT2::Drawing::IRainDrawer* rainDrawer, int32_t left, int32_t top, int32_t width, int32_t height);
+        using DrawSnowFunc = void (*)(
```
You only need one prototype for the weather drawing. Rename DrawRainFunc to DrawWeatherFunc

ZehMatt commented on src/openrct2/drawing/X8DrawingEngine.cpp:
```diff
@@ -224,7 +234,14 @@ void X8DrawingEngine::UpdateWindows()
 void X8DrawingEngine::PaintRain()
 {
-    DrawRain(&_bitsDPI, &_rainDrawer);
+    if (climate_is_snowing())
```
Same as above, move this out into its own.

ZehMatt commented on src/openrct2-ui/drawing/engines/opengl/OpenGLDrawingEngine.cpp:
```diff
@@ -342,7 +352,14 @@ class OpenGLDrawingEngine : public IDrawingEngine
     void PaintRain() override
```
```suggestion
    void PaintWeather() override
```
This is no longer only responsible for rain.

Gymnasiast commented:
It is, yes.

ZehMatt commented on src/openrct2-ui/UiContext.cpp:
```diff
@@ -860,6 +873,87 @@ class UiContext final : public IUiContext
         }
     }
+
+    static void DrawSnowWindow(
```
This is a literal copy of DrawRainWindow, consider better abstraction.

ZehMatt commented on src/openrct2-ui/UiContext.cpp:
```diff
@@ -281,6 +281,19 @@ class UiContext final : public IUiContext
         }
     }
 
+    void DrawSnowAnimation(IRainDrawer* rainDrawer, rct_drawpixelinfo* dpi, DrawSnowFunc drawFunc) override
```
This is a literal copy of DrawRainAnimation, consider better abstraction.

brettpenzer123 commented:
Hey, @ZehMatt thanks for your comments. I'm going to go back and just change the majority of the rain to weather and just include snow in that.

Generalcamo commented:
Should this be marked as "Closes #2350"?

brettpenzer123 commented:
I think the draft pull request would be better for just the developers to comment more on my code rather than a discussion. 

I believe #2350 allows discussion away from this PR.

brettpenzer123 commented:
Does anyone want to look over this again?

IntelOrca commented:
I tried out the changes and it seems to work okay.

* Blizzard could just be a snow version of thunderstorm.
* How are snow climates saved to SV6? Does it break vanilla loading?

ZehMatt commented:
The code looks okay now. I've tested around a bit with the snow movement and came up with a somewhat better looking version:
```
static void DrawLightSnow(IWeatherDrawer* weatherDrawer, int32_t left, int32_t top, int32_t width, int32_t height)
{
    const uint32_t t = gScenarioTicks / 2;
    const double t2 = (double)gScenarioTicks / 2.0f;

    int32_t x_start = -(int32_t)t + 1 + (cos(1.0 + left + t2 / 10.0) * 3);
    int32_t y_start = t + 1;
    y_start = -y_start;
    x_start += left;
    y_start += top;
    weatherDrawer->Draw(left, top, width, height, x_start, y_start, SnowPattern);

    x_start = -(int32_t)t + 16 + (cos(left + t2 / 10.0) * 3);
    y_start = t + 16;
    y_start = -y_start;
    x_start += left;
    y_start += top;
    weatherDrawer->Draw(left, top, width, height, x_start, y_start, SnowPattern);
}
```
This adds some side movement and reduced the falling speed by 2 which makes it look a bit more like snow imho.
![2020-03-01_19-14-29](https://user-images.githubusercontent.com/5415177/75631177-e9bd5700-5bf0-11ea-9b45-6f3d861bba82.gif)

brettpenzer123 commented:
> The code looks okay now. I've tested around a bit with the snow movement and came up with a somewhat better looking version:
> 
> ```
> static void DrawLightSnow(IWeatherDrawer* weatherDrawer, int32_t left, int32_t top, int32_t width, int32_t height)
> {
>     const uint32_t t = gScenarioTicks / 2;
>     const double t2 = (double)gScenarioTicks / 2.0f;
> 
>     int32_t x_start = -(int32_t)t + 1 + (cos(1.0 + left + t2 / 10.0) * 3);
>     int32_t y_start = t + 1;
>     y_start = -y_start;
>     x_start += left;
>     y_start += top;
>     weatherDrawer->Draw(left, top, width, height, x_start, y_start, SnowPattern);
> 
>     x_start = -(int32_t)t + 16 + (cos(left + t2 / 10.0) * 3);
>     y_start = t + 16;
>     y_start = -y_start;
>     x_start += left;
>     y_start += top;
>     weatherDrawer->Draw(left, top, width, height, x_start, y_start, SnowPattern);
> }
> ```
> 
> This adds some side movement and reduced the falling speed by 2 which makes it look a bit more like snow imho.
> ![2020-03-01_19-14-29](https://user-images.githubusercontent.com/5415177/75631177-e9bd5700-5bf0-11ea-9b45-6f3d861bba82.gif)

Yeah, that definitely looks a lot better. I'll definitely change it to that and slightly alter the heavy version.  Although when the code is implemented I'm sure other contributors will also change the way it looks.

brettpenzer123 commented:
> I tried out the changes and it seems to work okay.
> 
> * Blizzard could just be a snow version of thunderstorm ("thundersnow").
> * How are snow climates saved to SV6? Does it break vanilla loading?

That's quite a good idea with the thundersnow.

I'll need to take a look into the whole saving and loading with the vanilla version of the game. Since `current_rain_level > current_weather_level /
next_rain_level > next_weather_level ` within SV6 importing and exporting.

Generalcamo commented:
You did a git rename of rain.h to weather.h, but you appear to not have done the same with rain.cpp, which was deleted and recreated as weather.cpp with some the additions.

brettpenzer123 commented:
> You did a git rename of rain.h to weather.h, but you appear to not have done the same with rain.cpp, which was deleted and recreated as weather.cpp with the additions.

Yeah, it won't cause any issues. Just deleting rain.h would have made IntelliSense go absolutely crazy. So renaming it just made more sense and changing any of the includes to Weather.h from Rain.h.

brettpenzer123 commented:
![ezgif com-gif-maker (1)](https://user-images.githubusercontent.com/10356800/75713810-72f08e80-5cc2-11ea-8343-e7a4a971eadd.gif)
I know this is very frequent but just for gif purposes sped up, it uses exactly the same timer as a thunderstorm to create what @IntelOrca suggested thundersnow which would be in place for the blizzard.

brettpenzer123 commented on src/openrct2/drawing/LightFX.cpp:
```diff
@@ -923,7 +923,7 @@ void lightfx_apply_palette_filter(uint8_t i, uint8_t* r, uint8_t* g, uint8_t* b)
     overExpose *= 255.0f;
 
-    float targetFogginess = (float)(gClimateCurrent.RainLevel) / 8.0f;
+    float targetFogginess = (float)(gClimateCurrent.WeatherLevel) / 8.0f;
```
Did I miss something?

Generalcamo replied:
Misread the code, sorry.

brettpenzer123 replied:
No problem. I was triple checking everything and thought I'm so confused right now... lol

brettpenzer123 commented:
> > I tried out the changes and it seems to work okay.
> > 
> > * Blizzard could just be a snow version of thunderstorm ("thundersnow").
> > * How are snow climates saved to SV6? Does it break vanilla loading?
> 
> That's quite a good idea with the thundersnow.
> 
> I'll need to take a look into the whole saving and loading with the vanilla version of the game. Since `current_rain_level > current_weather_level / next_rain_level > next_weather_level ` within SV6 importing and exporting.

> > I tried out the changes and it seems to work okay.
> > 
> > * Blizzard could just be a snow version of thunderstorm ("thundersnow").
> > * How are snow climates saved to SV6? Does it break vanilla loading?
> 
> That's quite a good idea with the thundersnow.
> 
> I'll need to take a look into the whole saving and loading with the vanilla version of the game. Since `current_rain_level > current_weather_level / next_rain_level > next_weather_level ` within SV6 importing and exporting.

I've just taken a look at loading a saved game with the changes in this PR. 

Upon loading when the weather is snow, heavy snow or blizzard it is changed to heavy rain but sadly the sprite for heavy rain does not appear even though the weather has affectively changed to heavy rain. It's as if the weather sprite is still snow but the vanilla version can't show it. 

It does correct itself eventually back into the vanilla climate, players just have to wait for it.

IntelOrca commented:
I guess knowing that it does not crash vanilla, we can therefore allow it prior to new save format.

Gymnasiast commented on src/openrct2/actions/SetCheatAction.hpp:
```diff
@@ -347,7 +347,7 @@ DEFINE_GAME_ACTION(SetCheatAction, GAME_COMMAND_CHEAT, GameActionResult)
                 return { { 0, 255 }, { 0, 0 } };
             case CheatType::ForceWeather:
-                return { { 0, 5 }, { 0, 0 } };
+                return { { 0, 8 }, { 0, 0 } };
```
I prefer a constant or `std::size` here.

Gymnasiast commented on src/openrct2/world/Climate.cpp:
```diff
@@ -26,6 +26,7 @@
 
 #include <algorithm>
 #include <iterator>
+#include <openrct2\audio\Audio.cpp>
```
```suggestion
#include <openrct2/audio/Audio.cpp>
```

Gymnasiast commented on src/openrct2/world/Climate.cpp:
```diff
@@ -43,7 +44,7 @@ struct WeatherTransition
 
 extern const WeatherTransition* ClimateTransitions[4];
-extern const WeatherState ClimateWeatherData[6];
+extern const WeatherState ClimateWeatherData[9];
```
Could you introduce a `constexpr` or use `std::size` here, instead of a hardcoded number?

brettpenzer123 commented:
> I have taken a look. The code seems largely OK to me. I have left a few line notes.

I'll have a look at the code above.

> Do I understand correctly that you have ultimately implemented all three kinds of snow (regular snow, heavy snow and blizzard)?

Yes, that would be correct.

brettpenzer123 commented:
As of now, the strings (Snow, Heavy Snow, and Blizzard) are broken since the stringid isn't the value within the language files since this draft PR is quite old. I'll leave that till last to fix as it may keep changing making unnecessary commits for myself.

Gymnasiast commented on src/openrct2/actions/SetCheatAction.hpp:
```diff
@@ -347,7 +348,7 @@ DEFINE_GAME_ACTION(SetCheatAction, GAME_COMMAND_CHEAT, GameActionResult)
                 return { { 0, 255 }, { 0, 0 } };
             case CheatType::ForceWeather:
-                return { { 0, 8 }, { 0, 0 } };
+                return { { 0, WEATHER_COUNT }, { 0, 0 } };
```
```suggestion
                return { { 0, WEATHER_COUNT - 1 }, { 0, 0 } };
```

There are 9 different values, with 0 being the lowest and 8 being the highest.

brettpenzer123 replied:
Thanks :)

brettpenzer123 replied:
Thanks. I do this, to begin with, must have undone at some point.

brettpenzer123 commented:
Is there anything that needs changing or would you consider this complete?

Gymnasiast commented:
I want to take it for a spin. Once I can't find anything to nag about, I'll merge it or push the "Approve" button.

brettpenzer123 commented:
> I want to take it for a spin. Once I can't find anything to nag about, I'll merge it or push the "Approve" button.

Ok. 📦

Gymnasiast commented:
I have rebased and tested the PR. My findings:

- Judging from visuals, there appears to be no difference between heavy snow and blizzard. (I'd still expect a park not to be open in a blizzard, so we should perhaps take it out.)
- It started snowing in June, which isn't very realistic.

Other than that, it's looking good so far. :+1:

tupaschoal commented:
> It started snowing in June, which isn't very realistic.
If you're in the northern hemisphere 😛

brettpenzer123 commented:
> I have rebased and tested the PR. My findings:
> 
> * Judging from visuals, there appears to be no difference between heavy snow and blizzard. (I'd still expect a park not to be open in a blizzard, so we should perhaps take it out.)
> * It started snowing in June, which isn't very realistic.
> 
> Other than that, it's looking good so far. 👍

For the visual perspective, the blizzard was initially taken out but I added it again due to @IntelOrca suggesting
> I tried out the changes and it seems to work okay.
Blizzard could just be a snow version of a thunderstorm ("thundersnow").

So it acts exactly like a thunderstorm but with heavy snow. I know that it seems very unrealistic for a park to be open during a blizzard but I don't think a park would be open during long downpours of rain or a thunderstorm. In that regard, we either roll with what Roller Coaster Tycoon considers realistic or the park is forever closed due to weather.

Oh and since we don't have all twelve months in-game and we can't determine what hemisphere the player is currently playing from then we have no choice but for it to be random. Though in England it rains and snows whenever it feels like it. 😄 

I believe there may need to be some form of poll as we have mixed opinions on the weather blizzard.

Gymnasiast commented:
If the blizzard is to stay, I do think it has to differ a bit more from heavy snow. I haven't got a big problem with it being in the game per se.

As for the timing bits: the climates in OpenRCT2 are very obviously based on those in the northern hemisphere - it rains far more often in March, April, September and October, and it's hotter during the summer months, with a higher chance of thunderstorms. So I do think that will need to be adjusted.

Ziscor commented:
Before this is considered for merging, may I make a purely aesthetic minor suggestion? I tested the latest build, and think single pixel dots of individual snow could look better than the lines they are currently, or making the sprite less 'tall'. Looks like slowed rain the way it is now. For reference from Pokemon:
![image](https://user-images.githubusercontent.com/18721353/78360181-02e06b80-75d4-11ea-9555-0844b97b7776.png)

Just an opinion of course. And adding a sound effect at some later point would be nice to make the feature seem more 'complete' compared to vanilla rain. Something for an SFX project though I imagine. Just wanted to point out a quiet blizzard felt off. Thanks for coming back to work on this feature!

brettpenzer123 commented:
I'll take a look at this over the weekend. Thanks for the comments 👍

tupaschoal commented:
I'm labeling this as WIP, given the last comment (it also needs rebase)

brettpenzer123 commented:
> I'm labeling this as WIP, given the last comment (it also needs rebase)

Yeah. I haven't been able to work on it much. Partially because I've been having symptoms of covid-19 which wasn't fun at all, although feeling much better and that I've got a lot of catching up to do with my own projects.

tupaschoal commented:
It's alright, focus on getting healthy first :)

jessicajonesunitedkingdom commented:
hi

Is there any news from brett penzer?
Will he finish the snow so it can be added to the game?

I would really love it in the game!

thx

Gymnasiast commented:
I have decided to rebase this and take out the weather transitions for now (as they don't work correctly). I made a new PR for it because I wanted to keep the original branch as-is. The new PR is #12922.

I will keep this one open for a bit, until #12922 is merged.

Gymnasiast commented:
#12922 has now been merged, so closing this one too.

tupaschoal commented:
Has anything been left out that we should file an issue to track @Gymnasiast ? It felt like on the new pr text, but I can't be sure now

Gymnasiast commented:
@tupaschoal One thing has been left out, and that was the weather transitions. I made sure to put that into a separate commit: https://github.com/OpenRCT2/OpenRCT2/commit/ac985a8d61016b9afe33e70412c8d674d9a606ba

I'll create an issue for it.

tupaschoal commented:
Thanks!
